File: frontend/src/app/tasks/task.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import {
    CreateTaskRequest,
    PagedResult,
    TaskItem,
    TaskQueryParams,
    UpdateTaskRequest
} from './task.model';

@Injectable({ providedIn: 'root' })
export class TaskService {
    private readonly http = inject(HttpClient);
    private readonly baseUrl = `${environment.apiUrl}/tasks`;

    getTasks(query: TaskQueryParams = {}): Observable<PagedResult<TaskItem>> {
        let params = new HttpParams();
        if (query.status) params = params.set('status', query.status);
        if (query.priority) params = params.set('priority', query.priority);
        if (query.search) params = params.set('search', query.search);
        if (query.sortBy) params = params.set('sortBy', query.sortBy);
        if (query.sortDirection) params = params.set('sortDirection', query.sortDirection);
        if (query.page != null) params = params.set('page', String(query.page));
        if (query.pageSize != null) params = params.set('pageSize', String(query.pageSize));
        return this.http.get<PagedResult<TaskItem>>(this.baseUrl, { params });
    }

    getTask(id: string): Observable<TaskItem> {
        return this.http.get<TaskItem>(`${this.baseUrl}/${id}`);
    }

    createTask(request: CreateTaskRequest): Observable<TaskItem> {
        return this.http.post<TaskItem>(this.baseUrl, request);
    }

    updateTask(id: string, request: UpdateTaskRequest): Observable<TaskItem> {
        return this.http.put<TaskItem>(`${this.baseUrl}/${id}`, request);
    }

    deleteTask(id: string): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/${id}`);
    }
}

File: backend/src/TaskApi/Services/TaskService.cs
using Microsoft.EntityFrameworkCore;
using TaskApi.Data;
using TaskApi.Domain;
using TaskApi.Dtos;

namespace TaskApi.Services;

public sealed class TaskService(AppDbContext db) : ITaskService
{
    public async Task<PagedResult<TaskItemDto>> GetAllAsync(TaskQueryParams query, CancellationToken ct = default)
    {
        var q = db.Tasks.AsNoTracking();

        if (!string.IsNullOrWhiteSpace(query.Status) &&
            Enum.TryParse<Domain.TaskStatus>(query.Status, ignoreCase: true, out var statusEnum))
        {
            q = q.Where(t => t.Status == statusEnum);
        }

        if (!string.IsNullOrWhiteSpace(query.Priority) &&
            Enum.TryParse<TaskPriority>(query.Priority, ignoreCase: true, out var priorityEnum))
        {
            q = q.Where(t => t.Priority == priorityEnum);
        }

        if (!string.IsNullOrWhiteSpace(query.Search))
        {
            var search = query.Search.ToLower();
            q = q.Where(t => t.Title.ToLower().Contains(search) ||
                              (t.Description != null && t.Description.ToLower().Contains(search)) ||
                              (t.Assignee != null && t.Assignee.ToLower().Contains(search)));
        }

        q = (query.SortBy?.ToLower(), query.SortDirection?.ToLower()) switch
        {
            ("duedate", "asc")     => q.OrderBy(t => t.DueDate),
            ("duedate", _)         => q.OrderByDescending(t => t.DueDate),
            ("priority", "asc")    => q.OrderBy(t => t.Priority),
            ("priority", _)        => q.OrderByDescending(t => t.Priority),
            ("title", "asc")       => q.OrderBy(t => t.Title),
            ("title", _)           => q.OrderByDescending(t => t.Title),
            ("updatedat", "asc")   => q.OrderBy(t => t.UpdatedAt),
            ("updatedat", _)       => q.OrderByDescending(t => t.UpdatedAt),
            ("createdat", "asc")   => q.OrderBy(t => t.CreatedAt),
            _                      => q.OrderByDescending(t => t.CreatedAt)
        };

        var totalCount = await q.CountAsync(ct);

        var pageSize = Math.Clamp(query.PageSize, 1, 100);
        var page = Math.Max(query.Page, 1);
        var items = await q
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(t => t.ToDto())
            .ToListAsync(ct);

        return new PagedResult<TaskItemDto>(items, totalCount, page, pageSize);
    }

    public async Task<TaskItemDto?> GetByIdAsync(Guid id, CancellationToken ct = default)
    {
        var task = await db.Tasks.AsNoTracking().FirstOrDefaultAsync(t => t.Id == id, ct);
        return task?.ToDto();
    }

    public async Task<TaskItemDto> CreateAsync(CreateTaskRequest request, CancellationToken ct = default)
    {
        var status = Enum.TryParse<Domain.TaskStatus>(request.Status, ignoreCase: true, out var s)
            ? s
            : Domain.TaskStatus.Todo;

        var priority = Enum.TryParse<TaskPriority>(request.Priority, ignoreCase: true, out var p)
            ? p
            : TaskPriority.Medium;

        var now = DateTimeOffset.UtcNow;
        var task = new TaskItem
        {
            Id = Guid.NewGuid(),
            Title = request.Title,
            Description = request.Description,
            Status = status,
            Priority = priority,
            DueDate = request.DueDate,
            Assignee = request.Assignee,
            CreatedAt = now,
            UpdatedAt = now
        };

        db.Tasks.Add(task);
        await db.SaveChangesAsync(ct);

        return task.ToDto();
    }

    public async Task<TaskItemDto?> UpdateAsync(Guid id, UpdateTaskRequest request, CancellationToken ct = default)
    {
        var task = await db.Tasks.FirstOrDefaultAsync(t => t.Id == id, ct);
        if (task is null) return null;

        if (request.Title is not null)
            task.Title = request.Title;

        if (request.Description is not null)
            task.Description = request.Description;

        if (request.Status is not null &&
            Enum.TryParse<Domain.TaskStatus>(request.Status, ignoreCase: true, out var status))
            task.Status = status;

        if (request.Priority is not null &&
            Enum.TryParse<TaskPriority>(request.Priority, ignoreCase: true, out var priority))
            task.Priority = priority;

        if (request.DueDate.HasValue)
            task.DueDate = request.DueDate;

        if (request.Assignee is not null)
            task.Assignee = request.Assignee;

        task.UpdatedAt = DateTimeOffset.UtcNow;

        await db.SaveChangesAsync(ct);
        return task.ToDto();
    }

    public async Task<bool> DeleteAsync(Guid id, CancellationToken ct = default)
    {
        var task = await db.Tasks.FindAsync([id], ct);
        if (task is null) return false;

        db.Tasks.Remove(task);
        await db.SaveChangesAsync(ct);
        return true;
    }
}
